File: src/data/euHandlingTax.test.ts
import { describe, expect, it } from 'vitest'
import { calculateFinanceQuoteTax, normalizeFinanceTaxSettings, type FinanceTaxSettings } from './financeTaxSettings'
import { calculateFinanceQuoteFees, normalizeFinanceSurchargeSettings } from './financeSurchargeSettings'
import { type ChannelTaxRule } from './channelTaxRules'

const key = '1::燕文::C-test'
const rule = (amount: number, changes: Partial<ChannelTaxRule> = {}): ChannelTaxRule => ({ key, mode: 'fixed-order', amount, perKg: 0, currency: 'USD', ...changes })
function settings() {
  return normalizeFinanceTaxSettings({ countries: [
    { country: '欧盟', selected: true, enabled: true, fixedFeeUsd: 0, sortOrder: 1, channelRules: [rule(3)] },
    { country: '罗马尼亚', selected: true, enabled: true, fixedFeeUsd: 7, sortOrder: 2, euTaxMode: 'add-handling', channelRules: [rule(7)], handlingRules: [rule(1)] },
  ] })
}
const quote = (value: FinanceTaxSettings, country = '罗马尼亚', channelKey = key, weightKg = 0.5, eurUsd = 1.2) => calculateFinanceQuoteTax(value, country, channelKey.split('::')[1]!, 10, { channelKey, weightKg, usdCny: 6.7, eurUsd })
describe('EU duty plus member-country handling fee', () => {
  it.each(['罗马尼亚', 'Romania', 'RO'])('adds 3 + 1 for %s and preserves the original override', country => {
    const value = settings(), original = JSON.stringify(value)
    expect(quote(value, country)).toMatchObject({ taxUsd: 4, totalUsd: 14, fixedFeeUsd: 4, calculation: { rule: 'eu-handling-v1', euTaxUsd: 3, handlingFeeUsd: 1 } })
    expect(JSON.stringify(value)).toBe(original)
    expect(quote(value, '德国').taxUsd).toBe(3)
    expect(quote(value, '美国').taxUsd).toBe(0)
    value.countries.find(c => c.country === '罗马尼亚')!.euTaxMode = 'override'
    expect(quote(value).taxUsd).toBe(7)
  })
  it('round-trips mode and distinct fee rules without treating them as surcharge settings', () => {
    const value = normalizeFinanceTaxSettings(JSON.parse(JSON.stringify(settings())))
    expect(quote(value).taxUsd).toBe(4)
    expect(normalizeFinanceSurchargeSettings(value).countries.find(c => c.country === '罗马尼亚')).not.toHaveProperty('euTaxMode')
  })
  it.each(['no-tax', 'exempt'] as const)('EU %s does not waive processing fees', mode => {
    const value = settings()
    value.countries.find(c => c.country === '欧盟')!.channelRules = [rule(3, { mode })]
    expect(quote(value)).toMatchObject({ included: false, taxUsd: 1, calculation: { euTaxUsd: 0, handlingFeeUsd: 1 } })
  })
  it.each(['no-tax', 'exempt'] as const)('local %s does not waive EU duty', mode => {
    const value = settings()
    value.countries.find(c => c.country === '罗马尼亚')!.handlingRules = [rule(0, { mode })]
    expect(quote(value).taxUsd).toBe(3)
    value.countries.find(c => c.country === '罗马尼亚')!.handlingRules = []
    expect(quote(value).taxUsd).toBe(3)
  })
  it('blocks missing EU group and unavailable EU/local channels', () => {
    for (const name of ['欧盟', '罗马尼亚']) {
      const value = settings(), row = value.countries.find(c => c.country === name)!
      if (name === '欧盟') row.channelRules = [rule(0, { mode: 'unavailable' })]
      else row.handlingRules = [rule(0, { mode: 'unavailable' })]
      expect(quote(value)).toMatchObject({ configured: false, totalUsd: 10 })
    }
    const value = settings()
    value.countries.find(c => c.country === '欧盟')!.selected = false
    expect(quote(value).configured).toBe(false)
  })
  it('inherits legacy EU fixed duty, but blocks missing provider settings', () => {
    const value = settings(), eu = value.countries.find(c => c.country === '欧盟')!
    eu.channelRules = []; eu.fixedFeeUsd = 3; eu.providers = []
    expect(quote(value).configured).toBe(false)
    eu.providers = [{ provider: '燕文', mode: 'taxable', selected: true, channels: [] }]
    expect(quote(value).taxUsd).toBe(4)
  })
  it('adds processing once to the CHC EUR formula and converts its own currency', () => {
    const value = settings(), chc = '1::云途::C-600c364a09421e97a32f'
    value.countries.find(c => c.country === '罗马尼亚')!.handlingRules = [rule(6.7, { key: chc, currency: 'CNY' })]
    expect(quote(value, 'RO', chc)).toMatchObject({ taxUsd: 2.62, fixedFeeUsd: 0, feeMode: 'weight-order', calculation: { euTaxUsd: 1.62, handlingFeeUsd: 1 } })
    expect(quote(value, 'RO', chc, 1)).toMatchObject({ taxUsd: 3.52 })
    expect(quote(value, 'RO', chc, 0).configured).toBe(false)
    expect(quote(value, 'RO', chc, 1, 0).configured).toBe(false)
  })
  it('keeps additional fees separate and charges fixed handling once for every quantity', () => {
    const value = settings()
    const surcharges = normalizeFinanceSurchargeSettings({ countries: [{ country: '罗马尼亚', selected: true, enabled: true, sortOrder: 1, fixedFeeUsd: 2, exemptChannelKeys: [] }] })
    for (const quantity of [1, 2, 3, 5, 10]) {
      expect(calculateFinanceQuoteFees(value, surcharges, '罗马尼亚', '燕文', quantity * 10, key, { quantity, weightKg: quantity * 0.5, usdCny: 6.7 }))
        .toMatchObject({ taxUsd: 4, surchargeUsd: 2, totalUsd: quantity * 10 + 6 })
    }
  })
})

File: backend/src/test/java/com/milano/quotation/finance/EuHandlingTaxTest.java
package com.milano.quotation.finance;

import com.milano.quotation.common.AppException;
import org.junit.jupiter.api.Test;
import tools.jackson.databind.node.ObjectNode;
import java.math.BigDecimal;
import static org.junit.jupiter.api.Assertions.*;

class EuHandlingTaxTest {
    @Test void acceptsExplicitlyIncludedDutyAndHandlingButRejectsForgedInclusion() {
        var value=combinedSettings("exempt","exempt");
        var option=combinedOption(BigDecimal.ZERO,BigDecimal.ZERO,false).put("taxIncluded",true).put("taxFeeMode","exempt");
        assertTrue(ChannelTaxRules.validateQuote(value,option,exchange(),5));
        ((ObjectNode)value.path("countries").get(1).path("handlingRules").get(0)).put("mode","fixed-order");
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
    }
    private final ChannelTaxRulesTest fixtures = new ChannelTaxRulesTest();
    ObjectNode rule(String mode) { return fixtures.rule(mode); }
    ObjectNode settings(ObjectNode rule, String country) { return fixtures.settings(rule, country); }
    ObjectNode option(ObjectNode rule, BigDecimal amount) { return fixtures.option(rule, amount); }
    ObjectNode exchange() { return fixtures.exchange(); }
    ObjectNode combinedSettings(String euMode, String feeMode) {
        var value = settings(rule(euMode).put("amount",3), "欧盟");
        var local = value.withArray("countries").addObject().put("country","罗马尼亚").put("selected",true).put("enabled",true).put("euTaxMode","add-handling");
        local.putArray("channelRules").add(rule("fixed-order").put("amount",7));
        local.putArray("handlingRules").add(rule(feeMode).put("amount",1));
        return value;
    }
    ObjectNode combinedOption(BigDecimal eu, BigDecimal fee, boolean weighted) {
        var value = option(rule(weighted ? "weight" : "fixed-order"), eu.add(fee)).put("country","RO");
        for (var snapshot : value.path("taxCalculations")) {
            var node = (ObjectNode) snapshot;
            node.put("rule","eu-handling-v1").put("country","罗马尼亚").put("channelKey","1::燕文::C-1").put("euTaxUsd",eu).put("handlingFeeUsd",fee);
        }
        return value;
    }
    @Test void combinesCountryFeeOnceForAllQuantityColumnsAndRejectsEitherPriceChanging() {
        var value = combinedSettings("fixed-order","fixed-order");
        var option = combinedOption(new BigDecimal("3"),new BigDecimal("1"),false);
        assertTrue(ChannelTaxRules.validateQuote(value,option,exchange(),5));
        ((ObjectNode)value.path("countries").get(0).path("channelRules").get(0)).put("amount",4);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
        ((ObjectNode)value.path("countries").get(0).path("channelRules").get(0)).put("amount",3);
        ((ObjectNode)value.path("countries").get(1).path("handlingRules").get(0)).put("amount",2);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
    }
    @Test void exemptionsApplyOnlyToTheirOwnComponentAndUnsupportedRoutesAreBlocked() {
        for (var mode : new String[]{"no-tax","exempt"}) {
            assertTrue(ChannelTaxRules.validateQuote(combinedSettings(mode,"fixed-order"),combinedOption(BigDecimal.ZERO,BigDecimal.ONE,false),exchange(),5));
            assertTrue(ChannelTaxRules.validateQuote(combinedSettings("fixed-order",mode),combinedOption(new BigDecimal("3"),BigDecimal.ZERO,false),exchange(),5));
        }
        var option = combinedOption(new BigDecimal("3"),BigDecimal.ONE,false);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(combinedSettings("unavailable","fixed-order"),option,exchange(),5));
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(combinedSettings("fixed-order","unavailable"),option,exchange(),5));
    }
    @Test void currencyWeightAndSnapshotsAreRecomputed() {
        var value = combinedSettings("weight","fixed-order");
        ((ObjectNode)value.path("countries").get(0).path("channelRules").get(0)).put("amount",0.6).put("perKg",1.5).put("currency","EUR");
        ((ObjectNode)value.path("countries").get(1).path("handlingRules").get(0)).put("amount",6.7).put("currency","CNY");
        var option = combinedOption(new BigDecimal("2.01"),BigDecimal.ONE,true);
        assertTrue(ChannelTaxRules.validateQuote(value,option,exchange(),5));
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange().put("usdCny",7),5));
        ((ObjectNode)option.path("taxCalculations").path("5")).put("weightKg",0.1);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
    }
    @Test void removedEuOrChangedModeAndForgedTotalsRequireRepricing() {
        var value = combinedSettings("fixed-order","fixed-order");
        var option = combinedOption(new BigDecimal("3"),BigDecimal.ONE,false);
        ((ObjectNode)value.path("countries").get(0)).put("selected",false);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
        ((ObjectNode)value.path("countries").get(0)).put("selected",true);
        ((ObjectNode)value.path("countries").get(1)).put("euTaxMode","override");
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
        ((ObjectNode)value.path("countries").get(1)).put("euTaxMode","add-handling");
        option.put("taxCustomUsd",5);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
    }
    @Test void validatesNewSettingsAndPreservesChcDutyWhileAllowingItsSeparateHandling() {
        var local = (ObjectNode)combinedSettings("fixed-order","fixed-order").path("countries").get(1);
        ChannelTaxRules.validateSettings(local);
        local.put("country","美国");
        assertThrows(AppException.class,()->ChannelTaxRules.validateSettings(local));
        local.put("country","罗马尼亚");
        var fee=(ObjectNode)local.path("handlingRules").get(0);
        fee.put("key","1::云途::C-600c364a09421e97a32f");
        ChannelTaxRules.validateSettings(local);
        fee.put("amount",-1);
        assertThrows(AppException.class,()->ChannelTaxRules.validateSettings(local));
    }
    @Test void legacyEuDutyAndEmptyProcessingStillInheritAndRequireProviderConfiguration() {
        var value = combinedSettings("fixed-order","fixed-order");
        var eu = (ObjectNode)value.path("countries").get(0);
        eu.remove("channelRules"); eu.put("fixedFeeUsd",3);
        var option = combinedOption(new BigDecimal("3"),BigDecimal.ONE,false);
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange(),5));
        eu.putArray("providers").addObject().put("selected",true).put("provider","燕文").put("mode","taxable");
        assertTrue(ChannelTaxRules.validateQuote(value,option,exchange(),5));
        ((ObjectNode)value.path("countries").get(1)).remove("handlingRules");
        assertTrue(ChannelTaxRules.validateQuote(value,combinedOption(new BigDecimal("3"),BigDecimal.ZERO,false),exchange(),5));
    }
    @Test void builtInChcFormulaIsAddedToHandlingAndBlockedWithoutValidExchange() {
        var value = combinedSettings("fixed-order","fixed-order");
        var key = "1::云途::C-600c364a09421e97a32f";
        ((ObjectNode)value.path("countries").get(1).path("handlingRules").get(0)).put("key",key);
        var option = combinedOption(new BigDecimal("2.01"),BigDecimal.ONE,true).put("channelKey",key);
        for (var snapshot : option.path("taxCalculations")) ((ObjectNode)snapshot).put("channelKey",key);
        assertTrue(ChannelTaxRules.validateQuote(value,option,exchange(),5));
        assertThrows(AppException.class,()->ChannelTaxRules.validateQuote(value,option,exchange().put("eurUsd",0),5));
    }
}
